### PR TITLE
Cancellable async mutex - reimplementation

### PR DESCRIPTION
**Pull request by MaxMotovilov**
Reimplemented using an `std::mutex` and an intrusive double linked list instead of the lock-free queues used in the original implementation. As a simple optimization, the non-contended access to `async_mutex` is still lock-free.

Also changed the unintended behavior quirk of the original implementation by making `async_mutex` scheduler affine, following the pattern established by `async_manual_reset_event`.

### Test Plan:

```
make test
```
New unit tests were added to cover edge cases related to cancellation.

cc @ispeters

### CODE REVIEW

File: include/unifex/async_mutex.hpp [hunk 1]
```diff
@@ -78,68 +71,144 @@ class async_mutex {
...
+      mutex_.unlock();
+    } else {
+      mutex_.dequeue(this);
+    }
+    // Stop notification already happened, clean up here
+    stop_callback_.reset();
+    unifex::start(wrappedOp_);  // aka set_done(receiver)
+  } else if (!enqueued) {
+    set_value();
+  }
```
Comment (ispeters) [inline include/unifex/async_mutex.hpp:177]: I think this might work?
```suggestion
  stop_callback_.emplace(stop_token_, stop_callback{this});
  if (enqueued_state_.exchange(enqueued_state::kLocking, std::memory_order_acq_rel)
      == enqueued_state::kCancelled) {
    // stop callback ran synchronously
    stop_callback_.reset();
    completeWithDone();
    return;
  }
  
  // if this returns true then we've opened ourselves up to an unlocker
  // resuming us!
  bool enqueued{mutex_.try_enqueue(this)};
  
  auto oldState = enqueued_state::kLocking;
  auto newState = enqueued ? enqueued_state::kEnqueued
                           : enqueued_state::kLockedButNotEnqueued;
  
  if (enqueued_state_.compare_exchange_strong(
      oldState, newState, std::memory_order_acq_rel)) {
    // successfully transitioned locking -> enqueued|locked
    if (enqueued) {
      // we need to return here; the next state transition will happen when an unlocker
      // dequeues us or the stop request fires, both of which could be happening
      // concurrently to this code so it's unsafe to touch any members at this point
      return;
    } else {
      // we took the lock synchronously, so no unlocker will wake us up, but the stop
      // request could run
      stop_callback_.reset();
      if (auto state = enqueued_state_.load(std::memory_order_acquire);
          state == enqueued_state::kCancelled) {
        // the stop request ran
        completeWithDone();
      } else {
        assert(state == enqueued_state::kLockedButNotEnqueued);
        completeWithValue();
      }
      return;
    }
  } else {
    // we've been completed concurrently while trying to set up
    stop_request_.reset();
    
    if (oldState == enqueued_state::kCancelled) {
      // the stop request ran; we have to handle the possibility that we were enqueued onto
      // the list of waiters and then popped by an unlocker; we need to modify dequeue() to
      // return true if the item was in the list
      if (!enqueued || !mutex_.dequeue(this)) {
        // we got the lock but don't intend to keep it so give it back
        mutex_.unlock();
      }
      completeWithDone();
      return;
    } else {
      // we were enqueued but an unlocker dequeued us
      assert(oldState == enqueued_state::kUnlocked);
      assert(enqueued);
      completeWithValue();
      return;
    }
  }
```

We'd need the following definitions of `set_value` and `set_done`:
```cpp
void set_done() noexcept {
  auto oldState = enqueued_state_.load(std::memory_order_relaxed);

  do {
    if (oldState == enqueued_state::kUnlocked) {
      // the stop request ran after we acquired the lock
      return;
    }
  } while (!enqueued_state_.compare_exchange_weak(
      oldState, enqueued_state::kCancelled, std::memory_order_acq_rel));

  // we transitioned to cancelled; whether there's more to do depends on what from
  switch (oldState) {
    case enqueued_state::kUnlocked:
    default:
      // invalid state
      std::terminate();

    case enqueued_state::kNotEnqueued:
    case enqueued_state::kLocking:
    case enqueued_state::kLockedButNotEnqueued:
      // start() is still running
      return;

    case enqueued_state::kEnqueued:
      // start() put us in the queue
      if (!mutex_.dequeue(this)) {
        // an unlocker gave us the lock by popping us from the queue
        mutex_.unlock();
      }
      stop_callback_.reset();
      completeWithDone();
    }
}

void set_value() noexcept {
  // an unlocker has popped us from the mutex's queue
  auto oldState = enqueued_state_.load(std::memory_order_relaxed);

  do {
    if (oldState == enqueued_state::kCancelled) {
      // set_done is running; let it run
      return;
    }
  } while (!enqueued_state_.compare_exchange_weak(
      oldState, enqueued_state::kUnlocked, std::memory_order_acq_rel));

  // we transitioned to unlocked; whether there's more to do depends on what from
  if (oldState == enqueued_state::kLocking) {
    // start() is still running
    return;
  } else if (oldState == enqueued_state::kEnqueued) {
    // start() has returned or will soon and we beat the stop request to updating state
    stop_request_.reset();
    completeWithValue();
  } else {
    // invalid state
    std::terminate();
  }
}
```

**Review by ispeters — CHANGES_REQUESTED** [review]
I think there are latent deadlocks in the current implementation; I've taken a stab at describing an alternate implementation that might be free of deadlocks, but it needs validating.

File: include/unifex/async_mutex.hpp [hunk 2]
```diff
@@ -78,68 +71,144 @@ class async_mutex {
...
+    if (old_state == enqueued_state::kLockedButNotEnqueued) {
+      mutex_.unlock();
+    } else if (old_state == enqueued_state::kEnqueued) {
+      mutex_.dequeue(this);
+    } else {
+      // start() is not done yet, let it clean up
+      return;
+    }
+    stop_callback_.reset();
+    unifex::start(wrappedOp_);  // aka set_done(receiver)
```
[the remark this answers is not in the document]
Comment (MaxMotovilov) [reply include/unifex/async_mutex.hpp:201]: I think I borrowed this approach from some other unifex component - now I need to remember which one to decide if it was a bug to begin with!

File: include/unifex/async_mutex.hpp [hunk 3]
```diff
@@ -78,68 +71,144 @@ class async_mutex {
...
+    set_done() noexcept {
+  std::call_once(completed_, [this]() noexcept {
+    enqueued_state old_state{
+        enqueued_state_.exchange(enqueued_state::kCancelled)};
+    if (old_state == enqueued_state::kLockedButNotEnqueued) {
+      mutex_.unlock();
+    } else if (old_state == enqueued_state::kEnqueued) {
+      mutex_.dequeue(this);
+    } else {
+      // start() is not done yet, let it clean up
```
[the remark this answers is not in the document]
Comment (MaxMotovilov) [reply include/unifex/async_mutex.hpp:197]: Indeed - I think I got a bit taken with `call_once()` here and forgot that it only makes the method atomic in respect to the sender state, but not to mutex state! Back to the drawing board (which you have pre-filled already :-) ).

**Comment by MaxMotovilov** [comment]
The new version is being added as `unifex::v2::async_mutex`; old version stays as default (but moved into an inline namespace as `unifex::v1::async_mutex`). This should ensure no impact to existing users.
